File: python/deepnoodle/mobius/retry.py
"""Transient-response-aware retrying transport for the Mobius Python SDK.

Implements the shared retry policy documented in ``../docs/retries.md``.
Wraps an underlying :class:`httpx.BaseTransport` and transparently retries
eligible responses; surfaces a typed :class:`RateLimitError` when a 429
cannot be retried.
"""

from __future__ import annotations

import email.utils
import json
import time
from datetime import datetime, timezone
from typing import Callable

import httpx

from .errors import RateLimitError

DEFAULT_MAX_RETRIES = 3
MAX_RETRY_BACKOFF_SECONDS = 60.0
BASE_RETRY_BACKOFF_SECONDS = 1.0

_IDEMPOTENT_METHODS = frozenset({"GET", "HEAD", "PUT", "DELETE", "OPTIONS"})
_RETRYABLE_STATUS_CODES = frozenset({429, 500, 502, 503, 504})


class RetryingTransport(httpx.BaseTransport):
    """Retries transient responses and transport errors per the shared spec.

    Only GET/HEAD/PUT/DELETE/OPTIONS and replay-safe POST/PATCH requests —
    those carrying an ``Idempotency-Key`` header, or marked internally via
    ``request.extensions["replay_safe"]`` by the curated adopt-mode creates
    (whose idempotency comes from ``external_ref``) — are retried; other
    POST/PATCH requests surface ``RateLimitError`` immediately on 429 and
    re-raise transport errors immediately.
    """

    def __init__(
        self,
        base: httpx.BaseTransport,
        max_retries: int = DEFAULT_MAX_RETRIES,
        *,
        sleep: Callable[[float], None] = time.sleep,
        now: Callable[[], datetime] | None = None,
    ) -> None:
        self._base = base
        self._max_retries = max(0, int(max_retries))
        self._sleep = sleep
        self._now = now or (lambda: datetime.now(timezone.utc))

    def handle_request(self, request: httpx.Request) -> httpx.Response:
        attempt = 0
        idempotent = _is_idempotent(request)
        while True:
            response: httpx.Response | None = None
            try:
                response = self._base.handle_request(request)
                _prepare_replayable_json_response(request, response)
            except (httpx.TransportError, httpx.DecodingError):
                # No HTTP response was produced (connection reset, EOF,
                # I/O timeout, ...) or the acknowledgement body could not be
                # read and validated. Retry replayable, idempotent requests on
                # the same exponential-backoff budget; otherwise re-raise.
                if response is not None:
                    response.close()
                if not idempotent or attempt >= self._max_retries:
                    raise
                wait = _clamp(BASE_RETRY_BACKOFF_SECONDS * (2**attempt))
                if wait > 0:
                    self._sleep(wait)
                attempt += 1
                continue

            status = response.status_code

            if status not in _RETRYABLE_STATUS_CODES:
                return response

            out_of_budget = attempt >= self._max_retries or not idempotent

            if out_of_budget:
                if status == 429:
                    _drain(response)
                    raise _build_rate_limit_error(response)
                return response

            wait = _retry_after_or_backoff(response, attempt, self._now())
            _drain(response)
            if wait > 0:
                self._sleep(wait)
            attempt += 1

    def close(self) -> None:
        self._base.close()


def _is_idempotent(request: httpx.Request) -> bool:
    method = request.method.upper()
    if method in _IDEMPOTENT_METHODS:
        return True
    if method in {"POST", "PATCH"}:
        return _is_replay_safe_write(request)
    return False


def _is_replay_safe_write(request: httpx.Request) -> bool:
    """A POST/PATCH that opted into the replay-safe path.

    Either an ``Idempotency-Key`` header, or the internal per-request
    adopt-mode marker set by the curated create-or-adopt methods — never a
    public knob.
    """

    if request.headers.get("Idempotency-Key") not in (None, ""):
        return True
    return bool(request.extensions.get("replay_safe"))


def _prepare_replayable_json_response(
    request: httpx.Request,
    response: httpx.Response,
) -> None:
    """Complete replay-safe JSON acknowledgements inside the retry budget.

    SSE responses are deliberately left unread: once the response starts, a
    later stream failure must be handled by transcript reconnection rather
    than reinvoking the admission request.
    """

    content_type = response.headers.get("Content-Type", "").lower()
    if (
        request.method.upper() not in {"POST", "PATCH"}
        or not _is_replay_safe_write(request)
        or _accepts_event_stream(request.headers.get("Accept", ""))
        or not 200 <= response.status_code < 300
        or response.status_code == 204
        or "json" not in content_type
    ):
        return

    try:
        content = response.read()
        json.loads(content)
    except httpx.TransportError:
        raise
    except (json.JSONDecodeError, UnicodeDecodeError) as exc:
        raise httpx.DecodingError(
            "mobius: replay-safe response body is not valid JSON",
            request=request,
        ) from exc


def _accepts_event_stream(value: str) -> bool:
    return any(
        media_range.split(";", 1)[0].strip().lower() == "text/event-stream"
        for media_range in value.split(",")
    )


def _drain(response: httpx.Response) -> None:
    try:
        response.read()
    except httpx.StreamError:
        pass
    finally:
        response.close()


def _retry_after_or_backoff(
    response: httpx.Response,
    attempt: int,
    now: datetime,
) -> float:
    header = response.headers.get("Retry-After")
    parsed = _parse_retry_after(header, now)
    if parsed is not None:
        return _clamp(parsed)
    return _clamp(BASE_RETRY_BACKOFF_SECONDS * (2**attempt))


def _parse_retry_after(value: str | None, now: datetime) -> float | None:
    if value is None or value == "":
        return None
    stripped = value.strip()
    try:
        return float(int(stripped))
    except ValueError:
        pass
    parsed = email.utils.parsedate_to_datetime(stripped)
    if parsed is None:
        return None
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=timezone.utc)
    delta = (parsed - now).total_seconds()
    return max(0.0, delta)


def _clamp(seconds: float) -> float:
    if seconds < 0:
        return 0.0
    if seconds > MAX_RETRY_BACKOFF_SECONDS:
        return MAX_RETRY_BACKOFF_SECONDS
    return seconds


def _build_rate_limit_error(response: httpx.Response) -> RateLimitError:
    h = response.headers
    retry_after = _parse_retry_after(
        h.get("Retry-After"),
        datetime.now(timezone.utc),
    ) or 0.0
    limit = _int_or_none(h.get("X-RateLimit-Limit"))
    remaining = _int_or_none(h.get("X-RateLimit-Remaining"))
    reset_at: datetime | None = None
    reset_value = h.get("X-RateLimit-Reset")
    if reset_value:
        try:
            reset_at = datetime.fromtimestamp(int(reset_value), tz=timezone.utc)
        except (ValueError, OverflowError):
            reset_at = None
    scope = h.get("X-RateLimit-Scope") or None
    policy = h.get("X-RateLimit-Policy") or None
    return RateLimitError(
        retry_after=retry_after,
        limit=limit,
        remaining=remaining,
        reset_at=reset_at,
        scope=scope,
        policy=policy,
    )


def _int_or_none(v: str | None) -> int | None:
    if v is None or v == "":
        return None
    try:
        return int(v)
    except ValueError:
        return None

File: typescript/src/retry.ts
// 429/503- and transport-error-aware retrying `fetch` wrapper. Implements the
// shared retry policy documented in ../../docs/retries.md.

export class RateLimitError extends Error {
  readonly retryAfter: number;
  readonly limit: number | null;
  readonly remaining: number | null;
  readonly resetAt: Date | null;
  readonly scope: string | null;
  readonly policy: string | null;

  constructor(init: {
    retryAfter?: number;
    limit?: number | null;
    remaining?: number | null;
    resetAt?: Date | null;
    scope?: string | null;
    policy?: string | null;
    message?: string;
  } = {}) {
    const scope = init.scope ?? null;
    const scopeLabel = scope ?? "unknown";
    const retryAfter = init.retryAfter ?? 0;
    const message =
      init.message ??
      (retryAfter > 0
        ? `mobius: rate limit exceeded (scope=${scopeLabel}, retry after ${retryAfter}s)`
        : `mobius: rate limit exceeded (scope=${scopeLabel})`);
    super(message);
    this.name = "RateLimitError";
    this.retryAfter = retryAfter;
    this.limit = init.limit ?? null;
    this.remaining = init.remaining ?? null;
    this.resetAt = init.resetAt ?? null;
    this.scope = scope;
    this.policy = init.policy ?? null;
  }
}

export const DEFAULT_MAX_RETRIES = 3;
export const MAX_RETRY_BACKOFF_SECONDS = 60;
const BASE_RETRY_BACKOFF_SECONDS = 1;

const IDEMPOTENT_METHODS = new Set(["GET", "HEAD", "PUT", "DELETE", "OPTIONS"]);

type FetchFn = typeof globalThis.fetch;

export interface WrapRetryOptions {
  /** Number of retries for 429/503 responses. 0 disables retries. */
  maxRetries?: number;
  /** Override for tests — called in place of `setTimeout`-based sleeping. */
  sleep?: (seconds: number) => Promise<void>;
  /** Override for tests — called in place of `Date.now()` (milliseconds). */
  now?: () => number;
  /** Structured retry observation hook; never receives headers or bodies. */
  onRetry?: (event: RetryEvent) => void;
}

export interface RetryEvent {
  method: string;
  attempt: number;
  reason: "transport_error" | "rate_limited" | "service_unavailable";
  status?: number;
  waitSeconds: number;
}

/**
 * Wrap a `fetch` function so that 429 and 503 responses are retried per
 * the shared policy. The returned function is a drop-in replacement for
 * the global `fetch`.
 */
export function wrapFetchWithRetry(
  fetchFn: FetchFn,
  options: WrapRetryOptions = {},
): FetchFn {
  const maxRetries = Math.max(0, options.maxRetries ?? DEFAULT_MAX_RETRIES);
  const sleep = options.sleep ?? defaultSleep;
  const now = options.now ?? (() => Date.now());
  const onRetry = options.onRetry;

  return (async (input: RequestInfo | URL, init?: RequestInit) => {
    let attempt = 0;
    // Capture method / idempotency bits once — retries re-send the same request.
    const { method, hasIdempotencyKey, acceptsEventStream } = describeRequest(
      input,
      init,
    );
    const idempotent = isIdempotent(method, hasIdempotencyKey);

    while (true) {
      let response: Response | undefined;
      try {
        response = await fetchFn(input, init);
        // Fetch resolves once response headers arrive. For replay-safe JSON
        // writes, force the successful body to finish and parse before this
        // attempt counts as acknowledged. Reading a clone keeps the original
        // response available to the SDK's generated/high-level decoder.
        if (
          shouldValidateReplayableJSONResponse(
            response,
            method,
            hasIdempotencyKey,
            acceptsEventStream,
          )
        ) {
          await response.clone().json();
        }
      } catch (err) {
        if (response) await drainBody(response);
        // No HTTP response was produced (network error: connection reset,
        // DNS failure, I/O timeout, ...), or a successful replayable JSON
        // acknowledgement could not be fully read/parsed. Retry replayable,
        // idempotent requests on the exponential-backoff schedule; an aborted
        // request or an exhausted budget re-throws the underlying error.
        if (
          !idempotent ||
          attempt >= maxRetries ||
          isRequestCancelled(err, input, init)
        ) {
          throw err;
        }
        const wait = clamp(BASE_RETRY_BACKOFF_SECONDS * 2 ** attempt);
        onRetry?.({
          method,
          attempt: attempt + 1,
          reason: "transport_error",
          waitSeconds: wait,
        });
        if (wait > 0) {
          await sleep(wait);
        }
        attempt++;
        continue;
      }

      const status = response.status;
      if (status !== 429 && status !== 503) {
        return response;
      }

      const outOfBudget = attempt >= maxRetries || !idempotent;
      if (status === 429 && outOfBudget) {
        // Drain body so the connection can be reused.
        await drainBody(response);
        throw buildRateLimitError(response);
      }
      if (status === 503 && outOfBudget) {
        return response;
      }

      const wait = retryAfterOrBackoff(response, attempt, now);
      onRetry?.({
        method,
        attempt: attempt + 1,
        reason: status === 429 ? "rate_limited" : "service_unavailable",
        status,
        waitSeconds: wait,
      });
      await drainBody(response);
      if (wait > 0) {
        await sleep(wait);
      }
      attempt++;
    }
  }) as FetchFn;
}

function describeRequest(
  input: RequestInfo | URL,
  init?: RequestInit,
): {
  method: string;
  hasIdempotencyKey: boolean;
  acceptsEventStream: boolean;
} {
  let method = (init?.method ?? "GET").toUpperCase();
  let headers: Headers | undefined;
  if (init?.headers) {
    headers = new Headers(init.headers);
  }
  if (input instanceof Request) {
    if (!init?.method) method = input.method.toUpperCase();
    if (!headers) headers = new Headers(input.headers);
  }
  const hasIdempotencyKey = headers
    ? (headers.get("Idempotency-Key") ?? "").trim() !== ""
    : false;
  const acceptsEventStream =
    headers
      ?.get("Accept")
      ?.split(",")
      .some((value) => value.trim().toLowerCase() === "text/event-stream") ??
    false;
  return { method, hasIdempotencyKey, acceptsEventStream };
}

function isIdempotent(method: string, hasIdempotencyKey: boolean): boolean {
  if (IDEMPOTENT_METHODS.has(method)) return true;
  if (method === "POST" || method === "PATCH") return hasIdempotencyKey;
  return false;
}

function shouldValidateReplayableJSONResponse(
  response: Response,
  method: string,
  hasIdempotencyKey: boolean,
  acceptsEventStream: boolean,
): boolean {
  if (method !== "POST" && method !== "PATCH") return false;
  if (!hasIdempotencyKey || acceptsEventStream) return false;
  if (!response.ok || response.status === 204) return false;
  return (response.headers.get("Content-Type") ?? "")
    .toLowerCase()
    .includes("json");
}

// Caller cancellation and request timeouts are terminal. Checking the signal
// matters for AbortSignal.timeout(), whose rejection is named TimeoutError
// rather than AbortError and would otherwise retry with an already-aborted
// signal.
function isRequestCancelled(
  err: unknown,
  input: RequestInfo | URL,
  init?: RequestInit,
): boolean {
  const signal =
    init?.signal ?? (input instanceof Request ? input.signal : undefined);
  if (signal?.aborted) return true;
  if (typeof err !== "object" || err === null || !("name" in err)) {
    return false;
  }
  const name = (err as { name?: unknown }).name;
  return name === "AbortError" || name === "TimeoutError";
}

async function drainBody(response: Response): Promise<void> {
  try {
    await response.text();
  } catch {
    // ignore
  }
}

function retryAfterOrBackoff(
  response: Response,
  attempt: number,
  now: () => number,
): number {
  const header = response.headers.get("Retry-After");
  const parsed = parseRetryAfter(header, now);
  if (parsed !== null) return clamp(parsed);
  return clamp(BASE_RETRY_BACKOFF_SECONDS * 2 ** attempt);
}

function parseRetryAfter(value: string | null, now: () => number): number | null {
  if (value == null || value.trim() === "") return null;
  const trimmed = value.trim();
  if (/^-?\d+$/.test(trimmed)) {
    return Number(trimmed);
  }
  const ts = Date.parse(trimmed);
  if (Number.isNaN(ts)) return null;
  const delta = (ts - now()) / 1000;
  return Math.max(0, delta);
}

function clamp(seconds: number): number {
  if (!Number.isFinite(seconds) || seconds < 0) return 0;
  if (seconds > MAX_RETRY_BACKOFF_SECONDS) return MAX_RETRY_BACKOFF_SECONDS;
  return seconds;
}

function buildRateLimitError(response: Response): RateLimitError {
  const headers = response.headers;
  const retryAfter =
    parseRetryAfter(headers.get("Retry-After"), () => Date.now()) ?? 0;
  return new RateLimitError({
    retryAfter,
    limit: intOrNull(headers.get("X-RateLimit-Limit")),
    remaining: intOrNull(headers.get("X-RateLimit-Remaining")),
    resetAt: unixOrNull(headers.get("X-RateLimit-Reset")),
    scope: headers.get("X-RateLimit-Scope"),
    policy: headers.get("X-RateLimit-Policy"),
  });
}

function intOrNull(v: string | null): number | null {
  if (v == null || v === "") return null;
  const n = Number.parseInt(v, 10);
  return Number.isFinite(n) ? n : null;
}

function unixOrNull(v: string | null): Date | null {
  if (v == null || v === "") return null;
  const n = Number.parseInt(v, 10);
  return Number.isFinite(n) ? new Date(n * 1000) : null;
}

function defaultSleep(seconds: number): Promise<void> {
  return new Promise<void>((resolve) => setTimeout(resolve, seconds * 1000));
}
